### DISCUSSION TITLE
Transmission of the date and time in wifi

### DISCUSSION THREAD

**Discussion by PatLub85** [micropython/micropython#10646] [General]
Hi,
I have two esp8266 which communicate in wifi. 
The client sends its date and time with this piece of code:
```
def emission(donnees):
    adresse_serveur=socket.getaddrinfo("192.168.144.2",800)[0][-1]
    com_net=socket.socket(socket.AF_INET,socket.SOCK_STREAM)
    com_net.connect(adresse_serveur)
    com_net.send(donnees)
    com_net.close()

def hour():
   year = str(time.localtime()[0])
    month = str(time.localtime()[1])
    day = str(time.localtime()[2])
    hours = str(time.localtime()[3])
    mins = str(time.localtime()[4])
    if int(mins) < 10:
        mins = '0' + mins
    timestr = year + ':' + month + ':' +day +':' +hours + ':' + mins 
    return(timestr)

timestr=heure()
data=str(timestr)
print(data)
emission(data)
```
The line print (data) gives **2023:2:2:15:23**

The server with this piece of code:
```
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
s.bind((ip, 800))
s.listen(5)
conn, addr = s.accept()
reception_data=conn.recv(512)
      for item in reception_data.split(b':'):
          new_list.append(item.decode('ascii'))
print (new_list)
conn.close()
```
The line print (new_list) gives **['2000', '1', '1', '0', '00']** ???
Where is the problem ?

**Comment by davefes** [comment]
Try printing out `conn` and `addr` in your server code.
['2000', '1', '1', '0', '00'] is what a device running Micropython gives when booted-up and not connected to rshell or Thonny.

**Reply by karfas** [reply]
And, more important, print out the data you receive on the server side. 

Further I suspect that the code you posted isn't your server code, as micropython gives me a "IndentationError: unexpected indent" with an indented loop similar to yours.

**Comment by PatLub85** [comment]
Hi,
I should have started with that. I have an ESP8266_client which has a capacitive probe to measure soil humidity and a DHT22 to measure ambient air temperature and humidity.
After its deep-sleep, the ESP_client sends these three pieces of data, the server receives them, timestamps them, and sends them somewhere on the cloud.
On the other hand, if the connection is not made, the ESP_client saves the informations, timestamps them, and will send them back on its next connection.
This is the case I'm trying to make work.
Here is the simplified code for tests

sim_client.py
```
import network,socket
import machine
from machine import Pin
import time
from time import sleep
import dht

def config_wifi():
    ip        = '192.168.144.3'
    subnet    = '255.255.255.0'
    gateway   = '192.168.1.1'
    dns       = '8.8.8.8'
    ssid = 'MicroPython-AP'
    password = '123456789'
    sta=network.WLAN(network.STA_IF)
    sta.active(True)
    sta.ifconfig((ip,subnet,gateway,dns))
    sta.disconnect() # for ESP8266
    sta.connect(ssid,password)
    while (sta.isconnected()==False):
        sleep(1)
    ip=sta.ifconfig()[0]
    return ip

def emission(data):
    adress_server=socket.getaddrinfo("192.168.144.2",800)[0][-1]
    com_net=socket.socket(socket.AF_INET,socket.SOCK_STREAM)
    com_net.connect(adress_server)
    com_net.sendall(data.encode('uft8'))
    com_net.close()

def temp_hum():
    temp=20 #for example
    hum=50 #for example
    return (temp,hum)

def hour():
    year = str(time.localtime()[0])
    month = str(time.localtime()[1])
    day = str(time.localtime()[2])
    hour = str(time.localtime()[3])
    minu = str(time.localtime()[4])
    if int(minu) < 10:
        minu = '0' + minu
    timestr = year + ':' + month + ':' + day +':' +hour + ':' + minu
    return(timestr)

def hum_ground():
    variable = 80 #for example
    return(variable)

def run():
    config_wifi()
    while True:
        timestr=hour()
        variable=hum_ground()
        temp,hum = temp_hum()
        data=str(variable)+':'+str(temp)+':'+str(hum)+':'+timestr
        print(data)
        emission(data)

run()
```
sim_server.py
```
import socket
import network

ip        = '192.168.144.2'
subnet    = '255.255.255.0'
gateway   = '192.168.144.1'
dns       = '8.8.8.8'
ssid = 'MicroPython-AP'
password = '123456789'

sta=network.WLAN(network.STA_IF)
sta.active(True)
sta.disconnect() # pour ESP8266
ap = network.WLAN(network.AP_IF)
ap.active(True)
ap.ifconfig((ip,subnet,gateway,dns))
ap.config(essid=ssid, password=password)
while ap.active() == False:
  pass

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
s.bind((ip, 800))
s.listen(5)

def run():
    new_list=[]
    while True:
      conn, addr = s.accept()
      reception_data=conn.recv(512)
      for item in reception_data.split(b':'):
          new_list.append(item.decode('ascii'))
      print (new_list)
      conn.close()
  
run()
```

**Reply by davefes** [reply]
Did you try my suggestion?  Does the original problem remain?

I think there several errors in your server WiFi connect code.  IE, I do not see a connect and the last statement:
```py
while ap.active() == False:
  pass
```
doesn't look right

Read https://docs.micropython.org/en/latest/esp8266/quickref.html and check out the example code:
```py
def do_connect():
    import network
    wlan = network.WLAN(network.STA_IF)
    wlan.active(True)
    if not wlan.isconnected():
        print('connecting to network...')
        wlan.connect('ssid', 'key')
        while not wlan.isconnected():
            pass
    print('network config:', wlan.ifconfig())
```
Modified for your AP requirements.  The last print statement should give you confidence that you are connected in the mode that you want.

**Reply by bixb922** [reply]
How is the local time set? I don't see that code, for example ntptime.settime()

**Reply by PatLub85** [reply]
I use Thonny and it synchronizes the time. This is only for testing

**Reply by bixb922** [reply]
As @davefs wrote, it looks as if the client does not have the time set. 

Could you please print the data in the client just before sending, and then in the server just after receiving? I would also print the local time of the client to make sure it has the time set

**Reply by PatLub85** [reply]
Client:
function hour
print("Local time ：%s" %str(time.localtime())) --> Local time ：(2023, 2, 3, 19, 26, 41, 4, 34)

function emission
print('data ',data) --> data  80:20:50:2023:2:3:19:26
print('data encode',data.encode('utf-8')) --> data encode b'80:20:50:2023:2:3:19:26'

**Reply by PatLub85** [reply]
Server:
function run
print(reception_data) --> b'80:20:50:2000:1:1:0:00'
print (new_list) --> ['80', '20', '50', '2000', '1', '1', '0', '00']

**Reply by davefes** [reply]
So, the server is receiving (reception_data) not from your client.
```py
print str(add) --> ('192.168.144.3', 15422) the client'address
```
is not the address of your client which you set at 800

**Reply by davefes** [reply]
If you have several clients on 192.168.144.3 you might need to do a test on the port:
try;
```py
if (addr[1] == '800'):
    then
```

**Reply by bixb922** [reply]
You say the client sends  b'80:20:50:2023:2:3:19:26' and the server receives 
 b'80:20:50:2000:1:1:0:00'. That cannot be. 

Perhaps you have more than one client active?
Perhaps you are printing these values out at different times? 
Perhaps there is data queued in the send buffers?
Perhaps you have only one of the devices connected to Thonny at a time?
It's difficult to diagnose without further information about your setup.

**Reply by PatLub85** [reply]
Hi,
You ask the right questions!
How to look for problems where there are none!
I think I have found the solution to my problem, which in fact is not one.
Let me explain :
I just had the following experience:
The client on a USB port of my PC with Thonny, and the server on a second USB port with Putty. And there, the date sent and received is the same.
In fact yesterday I had the server on my PC with Thonny, and the client alone on a breadboard. Obviously, its internal clock was not updated, and it started from the year 2000.
The answer is quite simple.
Thank you all for helping me with this issue.

**Comment by PatLub85** [comment]
Hi Davefes,
Both connection codes are good. For example https://techtutorialsx.com/2017/06/06/esp32-esp8266-micropython-automatic-connection-to-wifi/
Here is the result of my print with my code : print('network config:', sta.ifconfig()) 
network config: ('192.168.144.3', '255.255.255.0', '192.168.1.1', '8.8.8.8')
I did not understand your suggestion. What do you mean?
And yes the problem is still there.
With my two simulation codes, it is possible to reproduce the problem

**Reply by davefes** [reply]
In your original server code add these two lines:
```py
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
s.bind((ip, 800))
s.listen(5)
conn, addr = s.accept()
reception_data=conn.recv(512)
      for item in reception_data.split(b':'):
          new_list.append(item.decode('ascii'))

print str(conn)
print str(add)

print (new_list)
conn.close()
``` 
Even better I suggest:
```py
print (f'conn = {conn}')
print (f'addr = {addr}')
print (f'new_list = {new_list}')
```
then you don't have to worry about what type you are wanting to print out.

**Reply by PatLub85** [reply]
print str(conn) --> <socket state=3 timeout=-1 incoming=0 off=0>
print str(add) --> ('192.168.144.3', 15422) the client'address

**Reply by davefes** [reply]
In your sim_server.py there is no connect().  

Do a:
```py
print('network config:', sta.ifconfig())
```
in sim_server.py

**Reply by PatLub85** [reply]
print str(conn) --> socket state=3 timeout=-1 incoming=0 off=0

**Reply by PatLub85** [reply]
The server is not in station mode, but in access point 
https://docs.micropython.org/en/latest/esp8266/tutorial/network_basics.html

**Reply by davefes** [reply]
So, a connection is not required.  Sorry, I have never use AP on either the ESP8266 or ESP32 as my servers are Linux boxes.
Suggestion:
- run your server code on a Linux box and then when you get that running change to the ESP8266 as a server.

**Reply by PatLub85** [reply]
Hi,
I don't have a Linux machine, only Windows11.
I know it's not right, but I'm an amateur !

For those whom it concerns print str(add) --> ('192.168.144.3', 15422), It's normal, the port that you see in the output is the source port and It's a random number.
See https://stackoverflow.com/questions/25026910/ports-with-python-sockets-are-they-random
